src/main/java/com/linecorp/armeria/common/ServiceInvocationContextAwareFuture.java

```diff
@@ -0,0 +1,132 @@
+package com.linecorp.armeria.common;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+
+final class ServiceInvocationContextAwareFuture<T> implements Future<T> {
+
+    private final ServiceInvocationContext context;
+    private final Future<T> delegate;
+
+    ServiceInvocationContextAwareFuture(ServiceInvocationContext context, Future<T> delegate) {
+        this.context = context;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public boolean isSuccess() {
+        return delegate.isSuccess();
+    }
+
+    @Override
+    public boolean isCancellable() {
+        return delegate.isCancellable();
+    }
+
+    @Override
+    public Throwable cause() {
+        return delegate.cause();
+    }
+
+    @Override
+    public Future<T> addListener(
+            GenericFutureListener<? extends Future<? super T>> listener) {
+        return delegate.addListener(context.makeContextAware(listener));
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    @Override
+    public Future<T> addListeners(
+            GenericFutureListener<? extends Future<? super T>>... listeners) {
+        return delegate.addListeners(Stream.of(listeners)
+                                           .map(context::makeContextAware)
+                                           .toArray(GenericFutureListener[]::new));
+    }
+
+    @Override
+    public Future<T> removeListener(
+            GenericFutureListener<? extends Future<? super T>> listener) {
+        return delegate.removeListener(listener);
```

Comment:
This will not work because we added a wrapped listener and a user will specify a unwrapped listener here. The default Promise implementation performs identity comparison, so overriding equals/hashCode() won't work as expected.